src/Parallel/LockfreeUnorderedSet.hpp

```diff
@@ -0,0 +1,290 @@
+   *
+   * \warning \p capacity must be a power of two greater than 0.
+   */
+  explicit LockfreeUnorderedSet(size_t capacity);
+  // Delete copy and move constructors and assignment operators since this
+  // class stores atomic variables needed for thread-safety.
+  LockfreeUnorderedSet(const LockfreeUnorderedSet&) = delete;
+  LockfreeUnorderedSet& operator=(const LockfreeUnorderedSet&) = delete;
+  LockfreeUnorderedSet(LockfreeUnorderedSet&&) = delete;
+  LockfreeUnorderedSet& operator=(LockfreeUnorderedSet&&) = delete;
+  ~LockfreeUnorderedSet() = default;
+
+  /*!
+   * \brief Insert the \p key into the set.
+   *
+   * \param key The key to insert.
+   * \param max_linear_probes The maximum number of linear probes to perform
+   *        before we give up. If `std::nullopt` (the default), then the maximum
+   *        number of linear probes is the container capacity (`capacity`).
+   *        This can be used to reduce the worst case linear probing cost from
+   *        O(capacity) to O(\p max_linear_probes) but users should be certain
+   *        this will not introduce bugs in their code.
+   * \return `true` if the key was inserted or found in the set. Returns
+   *         `false` if we failed to insert the key because we reached
+   *         \p max_linear_probes.
+   */
+  [[nodiscard]] bool insert(
+      T key, std::optional<size_t> max_linear_probes = std::nullopt) noexcept;
+
+  /*!
+   * \brief Erase the \p key from the set.
+   *
+   * \param key The key to erase.
+   * \param max_linear_probes The maximum number of linear probes to perform
+   *        before we give up. If `std::nullopt` (the default), then the maximum
+   *        number of linear probes is the container capacity (`capacity`).
+   *        This can be used to reduce the worst case linear probing cost from
+   *        O(capacity) to O(\p max_linear_probes) but users should be certain
+   *        this will not introduce bugs in their code.
+   * \return `true` if the key was erased by this thread. Returns `false` if we
+   *         failed to erase the key because we reached \p max_linear_probes
+   *         or because another thread erased the key. We have no general way of
+   *         knowing why we could not find the \p key to erase.
+   */
+  [[nodiscard]] bool erase(
+      T key, std::optional<size_t> max_linear_probes = std::nullopt) noexcept;
+
+  /*!
+   * \brief Check if the unordered set contains the \p key.
+   *
+   * \param key The key to check if it is contained in the unordered set.
+   * \param max_linear_probes The maximum number of linear probes to perform
+   *        before we give up. If `std::nullopt` (the default), then the maximum
+   *        number of linear probes is the container capacity (`capacity`).
+   *        This can be used to reduce the worst case linear probing cost from
+   *        O(capacity) to O(\p max_linear_probes) but users should be certain
+   *        this will not introduce bugs in their code.
```

Comment:
If you don't have a strong expectation that the value is actually present in the set, `contains` will have bad performance, since the `return false` path is always `O(max_linear_probes)`.  Not necessarily a problem, depending on use, but should be noted.